src/wallet/wallet.cpp

```diff
@@ -3038,69 +3001,89 @@ bool CWallet::CreateTransactionInternal(
                     return false;
                 }
 
-                nFeeNeeded = coin_selection_params.m_effective_feerate.GetFee(nBytes);
-                if (nFeeRet >= nFeeNeeded) {
-                    // Reduce fee to only the needed amount if possible. This
-                    // prevents potential overpayment in fees if the coins
-                    // selected to meet nFeeNeeded result in a transaction that
-                    // requires less fee than the prior iteration.
-
-                    // If we have no change and a big enough excess fee, then
-                    // try to construct transaction again only without picking
-                    // new inputs. We now know we only need the smaller fee
-                    // (because of reduced tx size) and so we should add a
-                    // change output. Only try this once.
-                    if (nChangePosInOut == -1 && nSubtractFeeFromAmount == 0 && pick_new_inputs) {
-                        unsigned int tx_size_with_change = nBytes + coin_selection_params.change_output_size + 2; // Add 2 as a buffer in case increasing # of outputs changes compact size
-                        CAmount fee_needed_with_change = coin_selection_params.m_effective_feerate.GetFee(tx_size_with_change);
-                        CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
-                        if (nFeeRet >= fee_needed_with_change + minimum_value_for_change) {
-                            pick_new_inputs = false;
-                            nFeeRet = fee_needed_with_change;
-                            continue;
-                        }
-                    }
```

Comment:
This code is needed because of the looping behavior. As we are removing that loop, it is no longer needed.

As for why this was needed, suppose in one iteration of the loop, we choose a lot of small inputs. This will require a higher transaction fee. If those inputs are not enough to cover the fee that they need, we iterate again. In the next iteration, we find that with the increased target, we need only a few large inputs. This thus requires a lower transaction fee than the previous iteration. Since in this second iteration we have selected enough inputs to cover the transaction, we would exit the loop. But because we have used the previous iteration to set the change and fee, we are actually overpaying the fees for this input set. So this code was added to increase the change and reduce the fee to just the necessary amount in this scenario.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/17331#discussion_r631405310

> As for why this was needed, suppose in one iteration of the loop, we choose a lot of small inputs. [...]

Sorry, I understood why change output *boosting* is needed. The code I was asking about was the code from 0f402b9263b0579b29aa0f841fc64ad58d3efba6 in #10712 and now removed here which tried to *add* a change output late in the loop. I don't see any case it would trigger because the change output would already be added earlier in the loop.

https://github.com/bitcoin/bitcoin/blob/0f402b9263b0579b29aa0f841fc64ad58d3efba6/src/wallet/wallet.cpp#L2683
https://github.com/bitcoin/bitcoin/blob/0f402b9263b0579b29aa0f841fc64ad58d3efba6/src/wallet/wallet.cpp#L2766

This is all after coin selection.

Reply:
It's change boosting but in the case where we don't have change.

Reply:
> It's change boosting but in the case where we don't have change.
> 
> It's the same scenario as change boosting, but instead of having selected more than enough to cover the fees from the previous iteration, we have selected exactly enough to cover those fees. But in this second iteration, there are fewer inputs, so fewer fees, so we need to add a change output because there is no existing change to boost.

I keep linking to line 2683: https://github.com/bitcoin/bitcoin/blob/0f402b9263b0579b29aa0f841fc64ad58d3efba6/src/wallet/wallet.cpp#L2683

Because in every case where the code below it (line 2766) would seemingly need to add a change output, it seems like this line would already have added it. This line happens after selecting coins in the same iteration of the while loop. My only understanding (which makes no sense) is that the code is selecting coins, seeing if a change output needs to be added, then adding it on line 2683. Then seeing if a change output needs to be added again on line 2766, but this code can never trigger because the change output should be already added.

If the code added in 0f402b9263b0579b29aa0f841fc64ad58d3efba6 is just nonsense that doesn't do anything, then it should be safe to drop. And it could be safe to drop anyway. Everything else in this PR looks good, just these CreateTransactionInternal changes are still pretty opaque to me.

Reply:
I think the key thing in understanding this code is that `nChange` (now named `change_and_fee`) was actually supposed to just be the change. `nValueToSelect` includes the `nFeeRet` from the previous iteration. So we could have `nChange == 0` if we select exactly enough to cover the recipient amount + previous `nFeeRet`. That would cause this code to trigger.

Reply:
Oh, wow. I guess I was looking at the new code first where `nChange` means input exceeding originally requested output, or change+fees. But in the current code as long as nValueToSelect still contains fees, then nChange will just be pure change:

https://github.com/bitcoin/bitcoin/blob/0f402b9263b0579b29aa0f841fc64ad58d3efba6/src/wallet/wallet.cpp#L2620
https://github.com/bitcoin/bitcoin/blob/0f402b9263b0579b29aa0f841fc64ad58d3efba6/src/wallet/wallet.cpp#L2664

So `nChange` could be 0 even if the final fee turned out to be a significantly less than the originally estimated fee, and the change output would not be added on the first pass.

I guess now the thing I don't understand is why commit "Move output reductions for fee to after coin selection" (234f85c44042d6566b7cf7e1bf1b969b2e38f903) is safe before commit "Remove CreateTransaction while loop and some related variables" (1f2cf3cf180a5928171729ce92324952cb172d9f) because it seems like the change output could still be missing until after the `nValueToSelect += nFeeRet;` line is dropped or no longer doing anything, resulting either in excessive fees paid again or just a failure (forgot details of the part below already....)

Reply:
Indeed, this commit fails the tests and doesn't quite work. However in the end, it all does work. I have tried to fix it, but it seems like it will end up just adding a bunch of code that gets removed in a later commit. I don't think there's a way to reorder commits to make this work.